src/main/java/org/codetracker/BlockTrackerImpl.java

```diff
@@ -0,0 +1,94 @@
+import org.codetracker.api.Version;
+import org.codetracker.change.Change;
+import org.codetracker.change.ChangeFactory;
+import org.codetracker.element.Block;
+import org.codetracker.element.Method;
+import org.codetracker.util.Util;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.lib.Repository;
+import org.refactoringminer.api.Refactoring;
+import org.refactoringminer.api.RefactoringType;
+import org.refactoringminer.rm1.GitHistoryRefactoringMinerImpl;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class BlockTrackerImpl extends BaseTracker implements BlockTracker {
+    private final ChangeHistory<Block> blockChangeHistory = new ChangeHistory<>();
+    private final String methodName;
+    private final int methodDeclarationLineNumber;
+    private final CodeElementType blockType;
+    private final int blockDeclarationLineNumber;
+    
+    public BlockTrackerImpl(Repository repository, String startCommitId, String filePath, String methodName, int methodDeclarationLineNumber, CodeElementType blockType, int blockDeclarationLineNumber) {
+        super(repository, startCommitId, filePath);
+        this.methodName = methodName;
+        this.methodDeclarationLineNumber = methodDeclarationLineNumber;
+        this.blockType = blockType;
+        this.blockDeclarationLineNumber = blockDeclarationLineNumber;
+    }
+
+    public static boolean checkOperationBodyChanged(OperationBody body1, OperationBody body2) {
+        if (body1 == null && body2 == null) return false;
+
+        if (body1 == null || body2 == null) {
+            return true;
+        }
+        return body1.getBodyHashCode() != body2.getBodyHashCode();
+    }
+
+    public static boolean checkOperationDocumentationChanged(VariableDeclarationContainer operation1, VariableDeclarationContainer operation2) {
+        String comments1 = Util.getSHA512(operation1.getComments().stream().map(UMLComment::getText).collect(Collectors.joining(";")));
+        String comments2 = Util.getSHA512(operation2.getComments().stream().map(UMLComment::getText).collect(Collectors.joining(";")));
+        return !comments1.equals(comments2);
+    }
+
+    private boolean isStartBlock(Block block) {
+        return block.getLocation().getCodeElementType().equals(blockType) &&
+        block.getLocation().getStartLine() <= blockDeclarationLineNumber &&
+                block.getLocation().getEndLine() >= blockDeclarationLineNumber;
+    }
+
+    private boolean isStartMethod(Method method) {
+        return method.getUmlOperation().getName().equals(methodName) &&
+                method.getUmlOperation().getLocationInfo().getStartLine() <= methodDeclarationLineNumber &&
+                method.getUmlOperation().getLocationInfo().getEndLine() >= methodDeclarationLineNumber;
+    }
+
+    @Override
+    public History<Block> track() throws Exception {
+        CompositeStatementObject composite;
```

Comment:
I'm not sure how I can get the relevant `composite` object in this context. Would appreciate any pointers.

Reply:
The main idea is to find the UMLOperationBodyMapper corresponding to the rightMethod (i.e., the current method from the child commit), and then get its mappings (a mapping is a pair of matched statements from left and right side).
If we get a mapping for `composite`, then we continue the tracking process with the left composite.

However, there are several optimizations we can do for performance.
If the container method has not changed, there is no need to run RefactoringMiner, or no need to run it for the entire class/file.